README
NOTE: Since Oliver Mai's homepage seems to be gone, I decided to
make this package available on the xfm website. Bob Woodside donated
the moxfm-1.0 sources and contributed some patches, and I gave moxfm
a little facelift using the 3dicons from the xfm and xplore packages.
Please note that this software is not maintained any longer, but I
hope you enjoy it anyway. Sep 03 2001 Albert Graef


					Hamburg 951214
_________________________________________________

MOXFM - the free Motif file manager (v. 1.0)
_________________________________________________


Moxfm was written using the source code of the
excellent Athena widgets based xfm file manager
and is intended to combine the advanced features
of xfm with the (IMHO) even more user-friendly
look and feel of the OSF/Motif (TM) GUI.

	(c) Simon Marlow 1990-1993 	simonm@dcs.glasgow.ac.uk
	(c) Albert Graef 1994, 1995	ag@muwiinfa.geschichte.uni-mainz.de
		(both for xfm)
	(c) Oliver Mai 1995, 1996	mai@sugra.desy.de


This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published
by the Free Software Foundation; either version 2, or (at your option) 
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

-------------------------------------------------------------------

DESCRIPTION

Moxfm is a full-fledged file and application manager for UN*X
platforms running the X Window system. It is quite similar to
the xfm file manager, whose source code moxfm is based upon,
and uses configuration files with the same format as xfm
(except the device configuration file, see the manpage moxfm(1)),
but there are quite some differences, too. If you are a xfm
user, you might want to read the README.moxfm-xfm file for the
most important differences between moxfm and xfm.

Moxfm - as xfm - allows you to browse your directory tree and to
copy, move, link and delete files in an intuitive way by simple
drag-and-drop actions. It provides context-sensitive popup menus.
Directory displays are updated automatically in regular intervals
when the contents of the directory changes.

The integrated application manager provides a kind of shelf on
which you can place your favorite applications as well as the
files and directories you are currently working with, and even
other (user-defined) types of dragged objects. It also
allows you to access different groups of applications and files.
User-definable file types let you specify a command to be
executed when double-clicking on a file or dropping other files
onto it. You can place files and applications as icons on the root
window and use them both for starting applications, opening
directory windows, and as drop sites. Application windows as well
as the desktop icons can also be updated automatically whenever the
corresponding application files change. Furthermore, moxfm provides
a monitoring facility which allows to ``have an eye'' on important
files and directories and respond with user-defined actions to
certain types of events when the status of one of these files
changes.

Last not least, moxfm can automatically mount and unmount
special devices like floppies as you open and close the
corresponding directories (mount points) or copy files to them.


INSTALLATION

Read the INSTALL file.
If you upgrade from an older version of moxfm, please read the
INSTALL file, too, as SOME FILE FORMATS HAVE CHANGED.

To compile, moxfm requires both OSF/Motif 1.2 or 2.0 and
Arnaud Le Hors' XPM library.
Binary versions of moxfm with Motif and the XPM library linked
statically are available for Linux-i386 (both ELF and a.out formats)
and HP-UX 9.x (see WHERE TO GET MOXFM).

If you are using the FVWM window manager, you should add the line
Style "desktop icon"	WindowListSkip
to your .fvwmrc (or to the system-wide system.fvwmrc).
You should not have the "NoPPosition" flag in your .fvwmrc in
order for desktop icon alignment to work.

If you are using a different window manager than fvwm or mwm,
the appearance of the desktop icons might not be as intended.
Twm doesn't remove the title bar, and ol[v]wm has some
problems, too. If you find a patch to solve these incompatibilities,
please tell me.

ICONS

Moxfm comes with a basic collection of pixmap icons from
various sources on the net, including icons from Linux
slackware, the X desktop manager (xdtm), the 0.91 beta
release of the G.R.E.A.T. environment, the AIcons collection
and the fvwm window manager. Most icons are taken from the
xfm-1.3.2 distribution. The icons with a beveled-edge
background where taken from an icon collection by Joerg
Mertin. If you want to use more of them, download
ftp://sunsite.unc.edu/pub/Linux/X11/icons/3d_pixmaps_1.0.tar.gz .
You might then want to change the Moxfm.appIconWidth and
Moxfm.appIconHeight resources to 56 and 46, respectively,
and change the background color, too.
There are no two-color bitmaps distributed with moxfm, but
they can be used as well, if you are running a version moxfm
linked with X11R6 and Motif 2.0.

A comprehensive set of pixmap fileicons has been collected
and edited for use with xfm by Juan D. Martin (juando@cnm.us.es).
These pixmaps are included in the moxfm distribution and are
installed during the 'make install' step.

Tons of other icons can be found on the net. For instance,
take a look at the AIcons collection or pixmaps.tar.gz, both
in the R5contrib directory on ftp.x.org. You might also consider
to install the xbmbrowser program written by Ashley Roll and
Anthony Thyssen (also on ftp.x.org), if you haven't done
this already.


STANDARD SETUP

The standard setup (files in the lib directory) consists of
the magic and xfmrc files for common filetypes, a sample moxfmdev
file (specifying how special devices like floppy disks are mounted
and unmounted), the targets configuration file (see the manpage
about it), and a set of useful application groups. It is
intended as a starting point for developing a personal setup which
fits your taste and requirements.

Quite some preconfigured application buttons require the xless browser,
which is available from ftp://ftp.x.org.
(Hint: xless looks much nicer, when you link it with the 3D version
of the Athena widgets, and use a non-white background color, e.g.
grey. Set the *shapeStyle resource to "rectangle"). A nicer
browser is pipescript from the Andrew toolkit (AUIS), but it takes
much more time to come up.


WHERE TO GET MOXFM

You can get the moxfm sources as well as binary distributions for
Linux (i386 architectures), HP-UX 9.x and IRIX 5.x from my WWW page
	http://ips105.desy.de:8765/~mai/moxfm
Both the sources and Linux binaries are also available via ftp from
	ftp://ftp.uni-stuttgart.de/pub/unix/systems/linux/local/X11/
A source-only distribution can be obtained from
	ftp://ftp.x.org/contrib/applications/moxfm-src.tgz

The xfm-1.3.2 sources, which moxfm is based on, are available at
ftp://ftp.x.org/contrib/applications/xfm-1.3.2.tar.gz


BUG REPORTS

Please send all bug reports and patches to mai@sugra.desy.de.
Always include the operation system of the machine you are
running moxfm on, as well as the Motif version.

Have fun!

Oliver Mai
mai@sugra.desy.de


ChangeLog
2001-09-03  root  <Dr.Graef@t-online.de>

	* changed program info (version number, www pointer)

	* fixed a bug in FmFw.c (eliminate trailing newline from ctime
	return value)

	* moxfm config files now go to .moxfm

	* face-lifting using modern apps and 3dicons, also tweaked some of
	the resources

	* added magic.c patches (from xplore/xfm)

	* added Bob Woodside's 1.0.1 patches (see below)

2001-08-15:	Version 1.0.1:

		New features:
                None.

                Changes:
                This is a minimal maintenance release designed to correct a
                few problems that have cropped up since Oliver left off work
                on moxfm, and produce a viable current distribution of the
                source.

                Bug Fixes:
                - regexp/regexp.c: Commented out argument-less declaration for
                strchr(), which causes compile failures with recent compilers.
                I know of no systems on which this declaration is needed any
                longer.
                - regexp/regsub.c: Commented out argument-less declaration for
                strncpy().
                - src/Common.h: Added logic to force define for SYSV on recent
                Linux systems, so that getwd will be properly redefined as
                getcwd. getwd is a dangerous function that generally should not
                be used, and recent versions of gcc issue a warning to this
                effect.
                - src/FmFw.c, src/FmIcons.c: Added a workaround for a failure
                of double-clicks to work under certain window managers/window
                manager modes, due to ill-synchronized passive grabs by the
                window manager on a parent window. This "fix" is based on a
                similar one made by Till Straumann for xfm.

**************
OLD ChangeLog:
**************

Date (yy-mm-dd):

95-08-17:	Moxfm v. 0.9 is available for public
		use for the first time via WWW. It is
		based on xfm v. 1.3.2.

95-08-21:	Version 0.9pl1: Bugfixes concerning 
		the state of the mount table window at
		startup 

95-08-22:	Version 0.91: Changes in function
		fnmultimatch() to allow for escaping
		blanks.

95-08-24:	Version 0.92: Changes in file window
		update code.

95-08-25:	Version 0.93: Code for updating file
		windows changed again. Mount point
		detection improved.

95-08-26:	Version 0.94: When copying, moving,
		linking or deleting fails for a number
		of files, the screen is no longer flooded
		with error message dialogs, but the
		user has the chance to cancel the
		operation. Changes in the code for
		recursive copying. Bugfixes.

95-08-28:	Version 0.94pl1: Bugfix concerning
		status display in connection with file
		selections.

95-08-29:	Version 0.94pl2: Bugfix in filesystem
		info dialog.

95-08-30:	Version 0.95: Changes in FmDragDrop.c
		in order to make moxfm compatible
		with future versions which (hopefully)
		will support copying files to/from a
		remote machine. Bugfix concerning
		-dir command line argument.

95-08-31:	Version 0.95pl1: Fixes in device mounting
		in FmFwActions.c and FmOps.c.

95-09-02:	Version 0.96: Major code clean-up of
		FmFwActions.c, also affecting other
		modules. Bugfixes here and in FmOps.c.

95-09-08:	Version 0.97: FIRST VERSION UPLOADED TO
		FTP-SERVER.
		
		- You cannot drop a file onto itself
		  any longer.
		
		Bugfixes:
		- When a filename is specified for an
		  application button, the filename is
		  passed as first argument also for
		  drop actions. This implies e.g. that
		  the drop action for the trashcan should
		  be
		  "shift;mv -f -b -V numbered $* ~/.trash"
		  now (as for xfm).
		- The icon names are set correctly now.

95-09-09:	Bugfix in usermount utility: userfstab
		file is checked for all mount options,
		not only "read-only" now.

95-09-18:	Version 0.98:

		New features:
		- There is a popup menu of the directory
		  name text field now (thanks to Steve
		  Blake for idea.)
		- The "File" pulldown menu has a "Perform
		  action" item now.
		- The mount table window has a "Close"
		  button.
		- Links are always made to *relative*
		  paths.
		- New resources saveOnExit and newWinOnDirPopup.
		
		Bugfixes:
		- Moxfm should hopefully compile on OSF/1
		  and SVR4 systems now (thanks to Albert
		  Graef and Volker Mertens).
		- Moxfm compiles with Motif 2.0 now.
		- Moxfm doesn't crash any longer when it
		  cannot allocate enough colors for the
		  pixmaps (thanks to Detlef Schmier).
		- The icon names should really be correct
		  now.
		  
		Changes:
		- When the autoSave feature of the
		  application windows is enabled, the
		  configuration is saved directly after
		  every change and not first when moxfm
		  exits.
		- The *relative* paths of dropped files
		  are passed to application buttons with
		  specified directories now.
		- There is a "Perform action" item in the
		  pulldown menu of directory icons, too.
		- New suppressWarnings resource instead of
		  suppressXtWarnings resource.
		- The names of the application group
		  configuration files don't start with "xfm-"
		  any longer.

96-05-16:	Version 1.0:

		New features:
		- Files and applications can be placed as icons
		on the root window.
		- Added automatic application window updates along
		with a checkApplicationFiles resource, and an applBoxIcon
		resource for specifying the standard application group
		icon. Furthermore, newly created application groups now
		point back to the group they were created in.
		(ag@muwiinfa.geschichte.uni-mainz.de)
		- Added a facility for monitoring files during
		automatic updates, along with the new xfmmon
		configuration file (AG).
		- You can drop user defined objects (listed in
		targets configuration file) on application windows.
		- You can switch off magic headers for certain
		devices and file windows.
		- You can move files across devices now.
		- The file type configuration file can be reread
		during runtime.

src/magic.c
/*-----------------------------------------------------------------------------
  Module magic.c

  (c) Juan D. Martin 1995
  (c) CNM-US 1995
                                                                           
  magic headers related code.
-----------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "magic.h"

/* ULTRIX apparently doesn't define these */
#ifdef ultrix
#define S_ISLNK(mode) (mode & S_IFMT) == S_IFLNK
#define S_ISSOCK(mode) (mode & S_IFMT) == S_IFSOCK
#endif

#define MAXLIN 1024
#define ASCLEN 512	
#define REGLEN 256	

static char    linebuf[MAXLIN];
static hmagic *mtypes = 0;	/* Array of file-types. */
static int     count = 0;	/* Number of types registered. */
static int     allocated = 0;	/* Room in the array. */
static int     top = 0;		/* Current top for sub-types. */
static int     maxhdr = ASCLEN;	/* Maximum header size required. */
static char   *hdrbuf = 0;	/* Header buffer allocated. */
static int     hdrbufsiz = 0;	/* Size of header buffer. */
static struct stat stbuf;	/* Stat buffer. */
static struct stat lstbuf;	/* Lstat buffer. */
static int bytes;		/* Bytes read from the file. */
static char   *namep;           /* pointer to filename */
static int     mmatch();
static char   *builtin_test();
static char   *builtin_result;
static char   *parse_string();

void magic_parse_file(name)
char *name;
{
    FILE *fh;
    int hsiz;

    count = 0; /* Added by O. Mai to allow reparsing of magic file */

    if(!hdrbuf)
    {
	hdrbuf = (char *) malloc(maxhdr + 1);
	hdrbufsiz = maxhdr;
    }

    if(!name || !(fh = fopen(name, "r")))
	return;

    while(fgets(linebuf, MAXLIN, fh))
    {
	char *cptr, *sptr;
	int cnt = 0;
	int l;

	l = strlen(linebuf);
	while(l < MAXLIN && linebuf[--l] == '\n' && linebuf[--l] == '\\')
	{
	    if(!fgets(linebuf + l, MAXLIN - l, fh))
		break;
	    l = strlen(linebuf);
	}

	if(count >= allocated)
	{
	    if(!mtypes)
	    {
		allocated = 25;
		mtypes = (hmagic *) malloc(allocated * sizeof(hmagic));
	    }
	    else
	    {
		allocated += 25;
		mtypes = (hmagic *) realloc((char *) mtypes,
					    allocated * sizeof(hmagic));
	    }
	}

	cptr = linebuf;
	if(cptr[0] == '>')	/* Sub-type */
	{
	    cptr++;
	    mtypes[top].subtypes++;
	}
	else if(cptr[0] == '#' || cptr[0] == '\n' || cptr[0] == '\r')
	{
	    continue;
	}
	else
	{
	    top = count;
	    mtypes[top].subtypes = 0;
	}

	mtypes[count].offset = strtol(cptr, &cptr, 0);
	while(isspace(*cptr))
	    cptr++;

	while(islower(cptr[cnt]))
	    cnt++;

	if(!strncmp("string", cptr, cnt))
	{
	    mtypes[count].flags = M_STRING;
	    cptr += cnt;

	    sptr = parse_string(&cptr);
	    mtypes[count].mask = strlen(sptr);
	    mtypes[count].value.string = 
		strcpy((char *)malloc(mtypes[count].mask + 1), sptr);

	    hsiz = mtypes[count].offset + mtypes[count].mask;
	    if(hsiz > maxhdr)
		maxhdr = hsiz;
	}
	else if(!strncmp("builtin", cptr, cnt))
	{
	    mtypes[count].flags = M_BUILTIN;
	    cptr += cnt;

	    sptr = parse_string(&cptr);
	    mtypes[count].value.string = 
		strcpy((char *)malloc(strlen(sptr) + 1), sptr);
	}
	else if(!strncmp("regexp", cptr, cnt))
	{
	    mtypes[count].flags = M_REGEXP;
	    cptr += cnt;


	    if(*cptr == '&')
	    {
		cptr++;
		mtypes[count].mask = strtol(cptr, &cptr, 0);
	    }
	    else
		mtypes[count].mask = REGLEN;
	    hsiz = mtypes[count].offset + mtypes[count].mask;
	    if(hsiz > maxhdr)
		maxhdr = hsiz;

	    sptr = parse_string(&cptr);
	    mtypes[count].value.expr = regcomp(sptr);
	}
	else
	{
	    long vmask;

	    if(!strncmp("byte", cptr, cnt))
	    {
		mtypes[count].flags = M_BYTE;
		hsiz = mtypes[count].offset + 1;
		if(hsiz > maxhdr)
		    maxhdr = hsiz;
		vmask = 0xFF;
	    }
	    else if(!strncmp("short", cptr, cnt))
	    {
		mtypes[count].flags = M_SHORT;
		hsiz = mtypes[count].offset + 2;
		if(hsiz > maxhdr)
		    maxhdr = hsiz;
		vmask = 0xFFFF;
	    }
	    else if(!strncmp("long", cptr, cnt))
	    {
		mtypes[count].flags = M_LONG;
		hsiz = mtypes[count].offset + 4;
		if(hsiz > maxhdr)
		    maxhdr = hsiz;
		vmask = 0xFFFFFFFF;
	    }
	    else if(!strncmp("mode", cptr, cnt))
	    {
		mtypes[count].flags = M_MODE;
		vmask = 0xFFFFFFFF;
	    }
	    else if(!strncmp("lmode", cptr, cnt))
	    {
		mtypes[count].flags = M_LMODE;
		vmask = 0xFFFFFFFF;
	    }
	    else
		continue;		/* Error. Skip line. */

	    cptr += cnt;

	    if(*cptr == '&')
	    {
		mtypes[count].flags |= M_MASKED;
		cptr++;
		mtypes[count].mask = strtol(cptr, &cptr, 0);
	    }

	    while(isspace(*cptr))
		cptr++;
	    switch(*cptr)
	    {
	    case '=':
		mtypes[count].flags |= M_EQ;
		cptr++;
		break;
	    case '<':
		mtypes[count].flags |= M_LT;
		cptr++;
		break;
	    case '>':
		mtypes[count].flags |= M_GT;
		cptr++;
		break;
	    case '&':
		mtypes[count].flags |= M_SET;
		cptr++;
		break;
	    case '^':
		mtypes[count].flags |= M_CLEAR;
		cptr++;
		break;
	    case 'x':
		mtypes[count].flags |= M_ANY;
		cptr++;
		break;
	    default:
		mtypes[count].flags |= M_EQ;
	    }
	    mtypes[count].value.number = strtol(cptr, &cptr, 0) & vmask;
	}
	while(isspace(*cptr))
	    cptr++;
	sptr = cptr;
	while(*cptr != '\n' && *cptr != '\r' && *cptr != '\0')
	    cptr++;
	*cptr = '\0';
	mtypes[count].message = strcpy((char *)malloc(strlen(sptr) + 1), sptr);
	count++;
    }
    fclose(fh);

    if(maxhdr > hdrbufsiz)
    {
	free(hdrbuf);
	hdrbuf = (char *) malloc(maxhdr + 1);
	hdrbufsiz = maxhdr;
    }
}

void magic_get_type(name, buf)
char *name;
char *buf;
{
    int i;
    int fd;

    namep = name;

    /* added hdrbuf initialization in case magic_get_type() is invoked
       before magic_parse_file() - AG */
    if(!hdrbuf)
    {
       hdrbuf = (char *) malloc(maxhdr + 1);
       hdrbufsiz = maxhdr;
    }

    if(lstat(name, &lstbuf) < 0)
    {
	stbuf.st_mode = 0;
	lstbuf.st_mode = 0;
	bytes = -1;
    }
    else
    {
	if(S_ISLNK(lstbuf.st_mode))
	{
	    if(stat(name, &stbuf) < 0)
	    {
		stbuf.st_mode = 0;
		bytes = -1;
	    }
	}
	else
	    stbuf = lstbuf;
	if(S_ISREG(stbuf.st_mode))
	{
	    if(stbuf.st_size == 0)
		bytes = 0;
            /* don't open zero-size files (causes hang on /proc) -- AG */
            else if((fd = open(name, O_RDONLY, 0)) < 0)
		bytes = -1;
	    else
	    {
		bytes = read(fd, hdrbuf, maxhdr);
		close(fd);
		/* Make sure it is nul-terminated. */
		if(bytes >= 0)
		    hdrbuf[bytes] = '\0'; 
	    }
	}
	else
	    bytes = 0;
    }

    builtin_result = (char *)0;
    for(i = 0; i < count; i++)
    {
	if(mmatch(i, buf))
	{
	    if(mtypes[i].subtypes != 0)
	    {
		char *nbuf;
		int n;

		if(buf[0] != '\0')
		    strcat(buf, " ");
		nbuf = buf + strlen(buf);
		n = i + mtypes[i].subtypes + 1;
		for(++i; i < n; i++)
		{
		    if(mmatch(i, nbuf) && nbuf[0] != '\0')
		    {
			strcat(nbuf, " ");
			nbuf = nbuf + strlen(nbuf);
		    }
		}
		n = strlen(buf) - 1;
		if(buf[n] == ' ')
		    buf[n] = '\0';
	    }
	    return;
	}
	else
	    i += mtypes[i].subtypes;
    }
    strcpy(buf, builtin_test());
}

static int mmatch(i, buf)
int i;
char *buf;
{
    int t;
    int o;
    long v;
    unsigned char *h;

    h = (unsigned char *) (hdrbuf + mtypes[i].offset);
    t = mtypes[i].flags & M_TYPE;

    switch(t)
    {
    case M_STRING:
	if(bytes >= mtypes[i].offset + mtypes[i].mask &&
	   !strncmp(hdrbuf + mtypes[i].offset, 
		    mtypes[i].value.string,
		    mtypes[i].mask))
	{
	    strcpy(buf, mtypes[i].message);
	    return 1;
	}
	return 0;

    case M_BUILTIN:
	if(!strcmp(mtypes[i].value.string, builtin_test()))
	{
	    strcpy(buf, mtypes[i].message);
	    return 1;
	}
	return 0;

    case M_REGEXP:
    {
	char sav;
	int len;
	regexp *prog = mtypes[i].value.expr;

	if(bytes <= mtypes[i].offset || !prog)
	    return 0;

	if(bytes > mtypes[i].offset + mtypes[i].mask)
	{
	    len = mtypes[i].mask;
	    sav = h[len];
	    h[len] = '\0';
	}
	else
	    len = -1;
	
	if(regexec(prog , h))
	{
	    regsub(prog, mtypes[i].message, buf);
	    if(len >= 0)
		h[len] = sav;
	    return 1;
	}
	if(len >= 0)
	    h[len] = sav;
    }
    return 0;

    case M_BYTE:
	if(bytes < mtypes[i].offset + 1)
	    return 0;
	else
	    v = h[0];
	break;

    case M_SHORT:
	if(bytes < mtypes[i].offset + 2)
	    return 0;
	else
	    v = ((long) h[0] << 8) | h[1];
	break;

    case M_LONG:
	if(bytes < mtypes[i].offset + 4)
	    return 0;
	else
	    v = ((long) h[0] << 24) | ((long) h[1] << 16) |
	        ((long) h[2] << 8) | h[3];
	break;

    case M_MODE:
	v = stbuf.st_mode;
	break;

    case M_LMODE:
	v = lstbuf.st_mode;
	break;

    default:
	return 0;		/* Should never happen. */
    }

    if (mtypes[i].flags & M_MASKED)
	v &= mtypes[i].mask;

    o = mtypes[i].flags & M_OP;
    switch(o)
    {
    case M_EQ:
	if(v == mtypes[i].value.number)
	{
	    sprintf(buf, mtypes[i].message, v);
	    return 1;
	}
	else
	    return 0;
    case M_LT:
	if(v < mtypes[i].value.number)
	{
	    sprintf(buf, mtypes[i].message, v);
	    return 1;
	}
	else
	    return 0;
    case M_GT:
	if(v > mtypes[i].value.number)
	{
	    sprintf(buf, mtypes[i].message, v);
	    return 1;
	}
	else
	    return 0;
    case M_SET:
	if((v & mtypes[i].value.number) == mtypes[i].value.number)
	{
	    sprintf(buf, mtypes[i].message, v);
	    return 1;
	}
	else
	    return 0;
    case M_CLEAR:
        /* AG fix: this must be "any clear" instead of "or" */
        if((~v & mtypes[i].value.number) != 0)
	{
	    sprintf(buf, mtypes[i].message, v);
	    return 1;
	}
	else
	    return 0;
    case M_ANY:
	sprintf(buf, mtypes[i].message, v);
	return 1;
    }
    return 0;
}

static char *builtin_test()
{
    int i;

    if(builtin_result)
	return builtin_result;
    if(bytes > 0)
    {
	if(bytes > ASCLEN)
	    bytes = ASCLEN;
	for(i = 0; i < bytes; i++)
	    if(!isascii(hdrbuf[i]))
		return builtin_result = "data";
	return builtin_result = "ascii";
    }
    if (bytes == 0)
    {
	if(S_ISREG(stbuf.st_mode))
	    return builtin_result = "empty";
	else
	    return builtin_result = "special";
    }
    return builtin_result = "unreadable";
}

static char *parse_string(s)
char **s;
{
    char cbuf[4];
    int i;
    char *vptr, *sptr, *cptr;

    cptr = *s;
    while(isspace(*cptr))
	    cptr++;
    vptr = sptr = cptr;

    while(!isspace(*cptr))
    {
	if(*cptr != '\\')
	{
	    *vptr++ = *cptr++;
	    continue;
	}
	cptr++;
	switch(*cptr)
	{
	case '\\':
	    *vptr++ = '\\';
	    cptr++;
	    break;
	case 'n':
	    *vptr++ = '\n';
	    cptr++;
	    break;
	case 'r':
	    *vptr++ = '\r';
	    cptr++;
	    break;
	case 't':
	    *vptr++ = '\t';
	    cptr++;
	    break;
	case 'v':
	    *vptr++ = '\v';
	    cptr++;
	    break;
	case 'b':
	    *vptr++ = '\b';
	    cptr++;
	    break;
	case 'f':
	    *vptr++ = '\f';
	    cptr++;
	    break;
	case 'x':
	    cptr++;
	    for(i = 0; i < 3 && isxdigit(*cptr); i++, cptr++)
		cbuf[i] = *cptr;
	    cbuf[i] = '\0';
	    *vptr++ = (char) strtol(cbuf, 0, 16);
	    break;
	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	    for(i = 0; i < 3 && *cptr >= '0' && *cptr <= '7';
		i++, cptr++)
		cbuf[i] = *cptr;
	    cbuf[i] = '\0';
	    *vptr++ = (char) strtol(cbuf, 0, 8);
	    break;
	default:
	    *vptr++ = *cptr++;
	}
    }
    *vptr = '\0';
    cptr++;
    *s = cptr;
    return sptr;
}
